File: src/main/kotlin/com/noahbres/meepmeep/roadrunner/trajectorysequence/sequencesegment/SequenceSegment.java
package com.noahbres.meepmeep.roadrunner.trajectorysequence.sequencesegment;

import com.acmerobotics.roadrunner.geometry.Pose2d;
import com.acmerobotics.roadrunner.trajectory.TrajectoryMarker;

import java.util.List;


/**
 * Abstract class representing a segment of a trajectory sequence.
 */
public abstract class SequenceSegment {
    /**
     * The duration of the segment.
     */
    private final double duration;

    /**
     * The starting pose of the segment.
     */
    private final Pose2d startPose;

    /**
     * The ending pose of the segment.
     */
    private final Pose2d endPose;

    /**
     * The list of trajectory markers associated with the segment.
     */
    private final List<TrajectoryMarker> markers;

    /**
     * Constructs a new SequenceSegment.
     *
     * @param duration  The duration of the segment.
     * @param startPose The starting pose of the segment.
     * @param endPose   The ending pose of the segment.
     * @param markers   The list of trajectory markers associated with the segment.
     */
    protected SequenceSegment(
            double duration,
            Pose2d startPose, Pose2d endPose,
            List<TrajectoryMarker> markers
    ) {
        this.duration = duration;
        this.startPose = startPose;
        this.endPose = endPose;
        this.markers = markers;
    }

    /**
     * Gets the duration of the segment.
     *
     * @return The duration of the segment.
     */
    public double getDuration() {
        return this.duration;
    }

    /**
     * Gets the starting pose of the segment.
     *
     * @return The starting pose of the segment.
     */
    public Pose2d getStartPose() {
        return startPose;
    }

    /**
     * Gets the ending pose of the segment.
     *
     * @return The ending pose of the segment.
     */
    public Pose2d getEndPose() {
        return endPose;
    }

    /**
     * Gets the list of trajectory markers associated with the segment.
     *
     * @return The list of trajectory markers.
     */
    public List<TrajectoryMarker> getMarkers() {
        return markers;
    }
}

File: src/main/kotlin/com/noahbres/meepmeep/roadrunner/trajectorysequence/SequenceSegment.kt
package com.noahbres.meepmeep.roadrunner.trajectorysequence

import com.acmerobotics.roadrunner.geometry.Pose2d
import com.acmerobotics.roadrunner.profile.MotionProfile
import com.acmerobotics.roadrunner.trajectory.Trajectory
import com.acmerobotics.roadrunner.trajectory.TrajectoryMarker

sealed class SequenceSegment(
        open val duration: Double,
        open val startPose: Pose2d,
        open val endPose: Pose2d,
        open val markers: List<TrajectoryMarker>
)

data class TrajectorySegment(
        val trajectory: Trajectory,
) : SequenceSegment(trajectory.duration(), trajectory.start(), trajectory.end(), emptyList())

data class TurnSegment(
        override val startPose: Pose2d,
        val totalRotation: Double,
        val motionProfile: MotionProfile,
        override val markers: List<TrajectoryMarker>
) : SequenceSegment(
        motionProfile.duration(),
        startPose,
        startPose.copy(heading = startPose.heading + totalRotation),
        markers
)

data class WaitSegment(
        private val pose: Pose2d,
        private val seconds: Double,
        override val markers: List<TrajectoryMarker>
) : SequenceSegment(seconds, pose, pose, markers)
